### PR TITLE
Inject relup instructions v4.4

### PR DESCRIPTION
**Pull request by terry-xiaoyu**
By using rebar3 3.14.3-emqx-8, we used to put the `dashboard.appup.src` as the last app to be upgraded in the generated relup file, to ensure `emqx_rule_engine:load_providers(), emqx_plugin:load().` is run after release upgrade is finished.

But this is not true with rebar3 from 3.18.0-emqx-1 (OTP24), as the feature has been removed from it.

We have to find another way to solve the problem.

This pull request injects additional instructions to the generated relup file for doing more works after release upgrade is finished.

The instruction injected:

For upgrade:

```
{apply, {emqx_relup,post_release_upgrade, [FromRelVsn, #{}]}
```

For downgrade:

```
{apply, {emqx_relup,post_release_downgrade, [ToRelVsn, #{}]}
```

### CODE REVIEW

File: scripts/inject-relup.escript [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            EmqxAppVsns = get_emqx_app_vsns(RelRootDir),
+            ok = inject_relup_file(RelRootDir, CurrRelVsn, EmqxAppVsns);
+        false ->
+            ?ERROR("not a valid root dir of release: ~p, for example: _build/emqx/rel/emqx",
+                [RelRootDir]),
+            erlang:halt(1)
+    end;
+main(_Args) ->
+    ?ERROR("~s", [usage()]),
+    erlang:halt(1).
```
Comment (zmstone) [inline scripts/inject-relup.escript:26]: do it like this:
```
stop() ->
    init:stop(1),
    timer:sleep(100),
    stop().
```
otherwise the io:format prints may race with halt.

File: bin/install_upgrade.escript [hunk 2]
```diff
@@ -144,9 +144,10 @@ parse_arguments([VersionStr|Rest], Acc) ->
     parse_arguments(Rest, [{version, Version}] ++ Acc).
 
 unpack_release(RelName, TargetNode, Version) ->
+    StartScriptExists = filelib:is_dir(filename:join(["releases", Version, "start.boot"])),
```
Comment (zmstone) [inline bin/install_upgrade.escript:147]: the changes in this file deserves a separate PR with a more focused description.

File: rebar.config.erl [hunk 3]
```diff
@@ -128,30 +128,34 @@ prod_compile_opts() ->
...
     [{add, [ {erl_opts, [deterministic]}]}].
 
-relup_deps(Profile) ->
-    {post_hooks, [{"(linux|darwin|solaris|freebsd|netbsd|openbsd)", compile, "scripts/inject-deps.escript " ++ atom_to_list(Profile)}]}.
+relup_deps(Profile, Vsn) ->
+    InjectCmd = "scripts/inject-relup.escript " ++ filename:join(["_build", Profile, "rel", "emqx"]) ++ " " ++ Vsn,
+    {post_hooks,
+        [ {"(linux|darwin|solaris|freebsd|netbsd|openbsd)", relup, InjectCmd}
+        ]
+    }.
```
Comment (zmstone) [inline rebar.config.erl:136]: this does not have to be a rebar hook.
add it to the `make_zip` function in the `build` script.
reason:
the relup file is already built and there is only one release to deal with.
i.e. so there is no need to do the wildcard below
`RelFiles = filelib:wildcard(filename:join([RelRootDir, "releases", "*", "emqx.rel"])),`

File: scripts/inject-relup.escript [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+-define(INFO(FORMAT, ARGS), io:format(user, "[inject-relup] " ++ FORMAT ++ "~n", ARGS)).
+
+usage() ->
+    "Usage: " ++ escript:script_name() ++ " <path-to-release-dir> <release-vsn>".
+
+main([RelRootDir, CurrRelVsn]) ->
+    case filelib:is_dir(filename:join([RelRootDir, "releases"])) andalso
+         filelib:is_dir(filename:join([RelRootDir, "lib"])) of
+        true ->
+            EmqxAppVsns = get_emqx_app_vsns(RelRootDir),
```
Comment (zmstone) [inline scripts/inject-relup.escript:17]: this is perhaps also no need for this versions collection,
as we aim to provide this feature for emqx app version 4.4.2
we can check if the upgrade to version is `>= 4.4.2` then there is a `emqx_relup` module, otherwise not.

File: scripts/inject-relup.escript [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    end, RUs).
+
+%% The `{apply, emqx_relup, post_release_upgrade, []}` will be appended to the end of
+%% the instruction lists.
+append_emqx_relup_instrs(up, EmqxAppVsns, CurrRelVsn, FromRelVsn, Instrs) ->
+    {EmqxVsn, true} = maps:get(CurrRelVsn, EmqxAppVsns),
+    Extra = #{}, %% we may need some extended args
+    %% we have to put 'load_object_code' before 'point_of_no_return'
+    %% so here we simply put it to the beginning
+    Instrs0 = [ {load_object_code, {emqx, EmqxVsn, [emqx_relup]}}
```
Comment (zmstone) [inline scripts/inject-relup.escript:67]: question: this instruction will be duplicated if the module is added to `emqx.appup.src`, is it ok to be loaded twice ?

if it's ok, please add `emqx_app` module here as well.
if not, we'll need to add a ci step to ensure both `emqx_relup` and `emqx_app` are always added to to `emqx.appup.src`.

**Review by zmstone — CHANGES_REQUESTED** [review]

File: scripts/inject-relup.escript (hunk 1 above)
Comment (terry-xiaoyu) [reply scripts/inject-relup.escript:26]: If the first `init:stop(1)` is running but not finished, will it work to try again?

Comment (terry-xiaoyu) [reply scripts/inject-relup.escript:26]: And I understand the `io:format/2` is an sync operation, is there really a race condition?

File: scripts/inject-relup.escript (hunk 4 above)
Comment (terry-xiaoyu) [reply scripts/inject-relup.escript:17]: Because the only thing I know is the ReleaseVsns when I am processing the relup file, I need the mapping from ReleaseVsns to EmqxAppVsns, e.g.:

```
#{ 4.4.0 => 4.4.1, %% the emqx vsn of release 4.4.0 is 4.4.1.
     4.4.1 => 4.4.2 %% the emqx vsn of release 4.4.1 is 4.4.2.
}
```
With the EmqxVsn in place, I can then construct the `load_object_code` instruction as:

```
{load_object_code, {emqx, EmqxVsn, [emqx_relup]}}
```

File: bin/install_upgrade.escript (hunk 2 above)
Comment (terry-xiaoyu) [reply bin/install_upgrade.escript:147]: I'll do this.

Comment (terry-xiaoyu) [reply bin/install_upgrade.escript:147]: https://github.com/emqx/emqx/pull/7165

File: rebar.config.erl (hunk 3 above)
Comment (terry-xiaoyu) [reply rebar.config.erl:136]: As I explained below, the collection for emqx vsn is necessary. So I don't know if we still need to put this script to `make_zip`.

File: scripts/inject-relup.escript (hunk 5 above)
Comment (terry-xiaoyu) [reply scripts/inject-relup.escript:67]: For duplicated instructions, it is OK if every `load` instruction has a predefined `load_object_code`. But if they are not appear in pairs, e.g. there're 2 `load_object_code` but 3 `load` after that, the release upgrade will crash. 

I'll put emqx_app here and also do more checks about the instructions.

File: scripts/inject-relup.escript (hunk 4 above)
Comment (zmstone) [reply scripts/inject-relup.escript:17]: If I understand it correctly, all we need to know is which app versions are ‘before’ emqc_relup module was added.
— it’s pretty static info which will not change in the future.

File: scripts/inject-relup.escript (hunk 1 above)
Comment (zmstone) [reply scripts/inject-relup.escript:26]: It happened to cuttlefish.
And it’s solved by doing the recursive stop.
https://github.com/emqx/cuttlefish/blob/6c346563e89ebbd95dbc1c29017adaf9abf85ca1/src/cuttlefish_escript.erl#L251

Comment (terry-xiaoyu) [reply scripts/inject-relup.escript:26]: That's probably it uses `logger` and `cuttlefish_test_logger`, which are async (by default).

I did the following experiment, it shows `io:format/2` always prints the msg before the vm is halt, but the logger can't:

```
{22-03-01 13:43}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> logger:error("hahah"), halt(1).

{22-03-01 13:43}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> logger:error("hahah"), halt(1).

{22-03-01 13:43}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> io:format("hahah"), halt(1).
hahah%

{22-03-01 13:44}MBP13:~/code/emqx44/_packages/emqx/emqx@inject_relup_instructions_v4.4 liuxy%
erl
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]
Eshell V12.1.5  (abort with ^G)
1> io:format("hahah"), halt(1).
hahah%
```

File: scripts/inject-relup.escript (hunk 4 above)
Comment (terry-xiaoyu) [reply scripts/inject-relup.escript:17]: So should we maintain a RelVsn -- EmqxVsn mapping file in the repo? So we don't need to hard code it in this script, nor we need to change this this script every time we release a new version.

Comment (zmstone) [reply scripts/inject-relup.escript:17]: I imagine this will be only applicable for 4.4.

If there’s a 4.5, emqx_relup is already there since day one.

File: build [hunk 6]
```diff
@@ -136,6 +136,7 @@ make_zip() {
     local tarball="${relpath}/${tarname}"
     local target_zip="${pkgpath}/${pkgname}"
     tar zxf "${tarball}" -C "${tard}/emqx"
+    scripts/inject-relup.escript "${tard}/emqx/releases/${PKG_VSN}/relup"
```
Comment (zmstone) [inline build:139]:
```suggestion
    ./scripts/inject-relup.escript "${tard}/emqx/releases/${PKG_VSN}/relup"
```

File: lib-ce/emqx_dashboard/src/emqx_dashboard.appup.src [hunk 7]
```diff
@@ -1,18 +1,11 @@
...
  ],
  [ {<<".*">>,
-    [ {apply, {emqx_rule_engine, load_providers, []}}
-    , {restart_application, emqx_dashboard}
-    , {apply, {emqx_plugins, load, []}}
+    [ {restart_application, emqx_dashboard}
     ]}
  ]
-}.
+}.
```
Comment (zmstone) [inline lib-ce/emqx_dashboard/src/emqx_dashboard.appup.src:11]:
```suggestion
}.
```

File: src/emqx.app.src [hunk 8]
```diff
@@ -6,7 +6,7 @@
   %% the emqx `release' version, which in turn is comprised of several
   %% apps, one of which is this.  See `emqx_release.hrl' for more
   %% info.
-  {vsn, "4.4.2"}, % strict semver, bump manually!
+  {vsn, "4.4.3"}, % strict semver, bump manually!
```
Comment (zmstone) [inline src/emqx.app.src:9]: there is no need to bump to 4.4.3 as 4.4.2 is not released yet.

File: src/emqx.appup.src [hunk 9]
```diff
@@ -1,7 +1,12 @@
 %% -*- mode: erlang -*-
 {VSN,
-  [{"4.4.1",
-    [{load_module,emqx_os_mon,brutal_purge,soft_purge,[]},
+  [{"4.4.2",
+    [{load_module,emqx_app,brutal_purge,soft_purge,[]},
+     {add_module,emqx_relup}]},
```
Comment (zmstone) [inline src/emqx.appup.src:5]: remove 4.4.2

File: src/emqx.appup.src [hunk 10]
```diff
@@ -30,8 +36,13 @@
      {load_module,emqx_message,brutal_purge,soft_purge,[]},
      {load_module,emqx_limiter,brutal_purge,soft_purge,[]}]},
    {<<".*">>,[]}],
-  [{"4.4.1",
-    [{load_module,emqx_os_mon,brutal_purge,soft_purge,[]},
+  [{"4.4.2",
+    [{load_module,emqx_app,brutal_purge,soft_purge,[]},
+     {delete_module,emqx_relup}]},
```
Comment (zmstone) [inline src/emqx.appup.src:41]: remove 4.4.2

File: scripts/inject-relup.escript [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+append_emqx_relup_instrs(down, ToRelVsn, Instrs0) ->
+    Extra = #{}, %% we may need some extended args
+    %% NOTE: When downgrading, we apply emqx_relup:post_release_downgrade/2 before reloading
+    %%  or removing the emqx_relup module.
+    Instrs1 = filter_and_check_instrs(down, Instrs0) ++
+        [ {load, {emqx_app, brutal_purge, soft_purge}}
+        , {apply, {emqx_relup, post_release_downgrade, [ToRelVsn, Extra]}}
+        ],
+    %% emqx_relup does not exist before release "4.4.1"
```
Comment (zmstone) [inline scripts/inject-relup.escript:67]:
```suggestion
    %% emqx_relup does not exist before release "4.4.2"
```

File: scripts/inject-relup.escript [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        ({load, {Mod, _, _}}, {EmqxAppVsn, EmqxMods, RemainInstrs})
+                when Mod =:= emqx_relup; Mod =:= emqx_app ->
+            {EmqxAppVsn, EmqxMods, RemainInstrs};
+        ({remove, {emqx_relup, _, _}}, {EmqxAppVsn, EmqxMods, RemainInstrs}) ->
+            {EmqxAppVsn, EmqxMods, RemainInstrs};
+        ({apply, {emqx_relup, _, _}}, {EmqxAppVsn, EmqxMods, RemainInstrs}) ->
+            {EmqxAppVsn, EmqxMods, RemainInstrs};
+        (Instr, {EmqxAppVsn, EmqxMods, RemainInstrs}) ->
+            {EmqxAppVsn, EmqxMods, RemainInstrs ++ [Instr]}
+    end, {"0.0.0", [], []}, Instrs).
```
Comment (zmstone) [inline scripts/inject-relup.escript:99]: nit: for better readability
```suggestion
    end, {not_found, [], []}, Instrs).
```

File: scripts/inject-relup.escript [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        case ToRelVsn of
+            "4.4.0" ++ _ -> [{remove, {emqx_relup, brutal_purge, brutal_purge}}];
+            "4.4.1" -> [{remove, {emqx_relup, brutal_purge, brutal_purge}}];
+            ToRelVsn -> [{load, {emqx_relup, brutal_purge, soft_purge}}]
+        end,
+    Instrs1 ++ LoadInsts.
+
+filter_and_check_instrs(Type, Instrs) ->
+    case take_emqx_vsn_and_modules(Instrs) of
+        {EmqxAppVsn, EmqxMods, RemainInstrs} when EmqxAppVsn =/= "0.0.0", EmqxMods =/= [] ->
```
Comment (zmstone) [inline scripts/inject-relup.escript:78]: nit:
```suggestion
        {EmqxAppVsn, EmqxMods, RemainInstrs} when EmqxAppVsn =/= not_found andalso EmqxMods =/= [] ->
```

**Review by zmstone — APPROVED** [review]
